Add TypeScript guides (#1960)
ember-learn/guides-source: guides/release/using-typescript/signatures.md

From a type checking perspective, these types must be _compatible_ with the types in the signature, though they do not have to be identical. Glint will type check that any types you write to make sure they are compatible. The rule for “compatibility” here is that your function signature types must be more general (“wider” in TypeScript terms) than the corresponding parts of the signature type.

For example, we could define the type of the positional arguments in the method body as `Array<unknown>` instead of `[string]`, while keeping the original signature’s `Positional: [string]`:

```typescript
  compute(positional: Array<unknown>, named: { locale?: string }): string {
    // ...
  }
```

Because the signature set on the class, callers would still have to pass a single string argument, but we would need to change the behavior of the body to [narrow the type][narrowing] for the first item in the array.

[narrowing]: http://www.typescriptlang.org/docs/handbook/2/narrowing.html

Accordingly, the best practice is to keep the types matching.

## Advanced signature techniques

> We can also define signatures in more complicated ways, using TypeScript features like generics and union types, to [make illegal states unrepresentable][illegal]. For example, a list component might need to yield back out instances of the same type it provides, and provide the appropriate element target based on a `type` argument.

Reviewer: ```suggestion
We can also define signatures in more complicated ways using more advanced TypeScript features.
Nearly anything you can do with a “regular” TypeScript function or class, you can also do with signatures for Glimmer invokables.
We can make a component accept a generic type, or use union types.
With these tools at our disposal, we can even define our signatures to [make illegal states unrepresentable][illegal].

To see this in practice, consider a list component which yields back out instances of the same type it provides, and provide the appropriate element target based on a `type` argument.
Yielding back out the same type passed in will use generics, and providing an appropriate element target for `...attributes` can use a union type.
```

    type: 'ordered';
  };
  Blocks: {
    default: [item: T];
  };
  Element: HTMLOListElement;
}

interface UnorderedList<T> {
  Args: {
    items: Array<T>;
    type: 'unordered';
  };
  Blocks: {
    default: [item: T];
  };
  Element: HTMLUListElement;
}

type ListSignature<T> = OrderedList<T> | UnorderedList<T>;

export default class List<T> extends Component<ListSignature<T>> {
  <template>
    {{#if (isOrdered @type)}}
      <ol ...attributes>
        {{#each @items as |item|}}
          <li>{{yield item}}</li>
        {{/each}}
      </ol>
    {{else}}
      <ul ...attributes>
        {{#each @items as |item|}}
          <li>{{yield item}}</li>
        {{/each}}
      </ul>
    {{/if}}
  </template>
}

function isOrdered(type: 'ordered' | 'unordered'): type is 'ordered' {
  return type === 'ordered';
}
```

When this component is invoked, the `@type` argument will determine what kinds of modifiers are legal to apply to it. For example, if you defined a modifier `reverse` which required an `HTMLOListElement`, this invocation would be rejected:

```handlebars
<List @items={{array 1 2 3}} @type='unordered' {{reverse}} as |item|>
  The item is
  {{item}}.
</List>
```

> 

Reviewer: ```suggestion

The same approach with generics works for class-based helpers and class-based modifiers.
Function-based helpers as and modifiers can also use generics, but by using them on the function definition rather than via a signature.
One caveat: particularly complicated union types in signatures can sometimes become too complex for Glint/TypeScript to resolve when invoking in a template.
In those cases, your best bet is to find a simpler way to structure the types while preserving safety.

```

}
```

We use the signature to provide both editor support for the invokable with TypeScript and [Glint][glint] and documentation using any tool which understands type annotations or JSDoc. Note that the `Blocks` and `Element` keys will only be used if you are type-checking your templates using [Glint][glint], but we recommend adding them now in order to be Glint-ready in the future.

[glint]: https://github.com/typed-ember/glint

A few things to note about these signatures:

> <!-- TODO: This was an ordered list, but nested list formatting is a bit funky currently, so I converted it to paragraphs. Would be nice to revert. -->

Reviewer: Honestly, the "First… Second… Third" is probably better anyway!

<button type='button' {{on 'click' this.pause}}>Pause</button>
```

> Let’s go one step further and switch to supporting for two [named blocks][nb]: a optional `title` block for a caption for the audio element, and a `fallback` block for the audio fallback where we previously used a `default` block.

Reviewer: Grammar nit:

```suggestion
Let’s go one step further and switch to supporting for two [named blocks][nb]: an optional `title` block for a caption for the audio element, and a `fallback` block for the audio fallback where we previously used a `default` block.
```
